viz/SonarVisualization.cpp

```diff
@@ -0,0 +1,176 @@
+#include "SonarVisualization.hpp"
+#include <osg/Geode>
+#include <base/Eigen.hpp>
+
+namespace vizkit3d
+{
+
+SonarVisualization::SonarVisualization():
+    motor_step(0),
+    num_steps(0),
+    new_sonar_scan(false),
+    full_scan(false)
+{
+    last_orientation.invalidate();
+    data.resize(1);
+}
+
+osg::ref_ptr< osg::Node > SonarVisualization::createMainNode()
+{
+    osg::ref_ptr<osg::Group> mainNode = new osg::Group();
+    pointGeom = new osg::Geometry;
+    pointsOSG = new osg::Vec3Array;
+    pointGeom->setVertexArray(pointsOSG);
+    color = new osg::Vec4Array;
+    pointGeom->setColorArray(color);
+    pointGeom->setColorBinding(osg::Geometry::BIND_PER_VERTEX);
+    pointGeom->getOrCreateStateSet()->setMode(GL_LIGHTING, osg::StateAttribute::OFF); 
+    drawArrays = new osg::DrawArrays( osg::PrimitiveSet::TRIANGLES, 0, pointsOSG->size() );
+    pointGeom->addPrimitiveSet(drawArrays.get());
+    
+    osg::ref_ptr<osg::Geode> geode = new osg::Geode;
+    geode->addDrawable(pointGeom.get());
+    mainNode->addChild(geode);
+    return mainNode;
+}
+
+/**
+ * Updates the class with the new sonar data of type SonarScan.
+ * 
+ * @param data new sonar data
+ */
+void SonarVisualization::updateDataIntern(const base::samples::Sonar& data)
+{
+    if(motor_step == 0)
+        return;
+    double current_pos = 0;
+    if(full_scan)
+        current_pos = ((data.bearings[0].rad+M_PI)/(2.0*M_PI))*num_steps;
+
+    SonarHelper bm;
+    bm.sonar_sample = data;
+    bm.orientation = last_orientation;
```

Comment:
The placement of the plugin should be done through the transformer, not some internal pose handling. Note that if you do want to manually place it, you can manage its frame within vizkit3d. My point is: don't transform internally to the plugin.